### PR TITLE
refactor(ir): make DataTypes annotable

### PR DESCRIPTION
**Pull request by kszucs**

### CODE REVIEW

File: ibis/common/caching.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        identifiers = tuple(id(item) for item in key)
+        value, _ = self._data[identifiers]
+        return value
+
+    def __delitem__(self, key):
+        identifiers = tuple(id(item) for item in key)
+        del self._data[identifiers]
+
+    def __repr__(self):
+        return repr(self._data)
```
Comment (cpcloud) [inline ibis/common/caching.py:49]: I know this was pre-existing from the previous PR, but can we show something indicating that this is not a plain `dict`?

```suggestion
        return f"{self.__class__.__name__}({self._data})"
```

File: ibis/common/grounds.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        params = {}
+        for parent in bases:
+            try:
+                signature = parent.__signature__
+            except AttributeError:
+                pass
+            else:
+                params.update(signature.parameters)
+        # store the originally inherited keys so we can reorder later
+        inherited = set(params.keys())
```
Comment (cpcloud) [inline ibis/common/grounds.py:85]: It should be fine to leave this as `params.keys()`.

File: ibis/tests/expr/test_value_exprs.py [hunk 3]
```diff
@@ -234,19 +234,19 @@ def test_literal_with_non_coercible_type(value, expected_type):
...
 
-    with pytest.raises(TypeError, match=expected_msg):
-        ibis.literal(value)
+#     with pytest.raises(TypeError, match=expected_msg):
+#         ibis.literal(value)
 
-    point = ibis.literal(value, type='point')
-    assert point.type() == dt.point
+#     point = ibis.literal(value, type='point')
+#     assert point.type() == dt.point
```
Comment (cpcloud) [inline ibis/tests/expr/test_value_exprs.py:249]: Can we uncomment these tests?

File: ibis/common/tests/test_grounds.py [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+    def __repr__(self):
+        return f"Node(name={self.name})"
+
+    def __equals__(self, other):
+        Node.num_equal_calls += 1
+        return self.name == other.name
+
+
+@pytest.fixture
+def cache():
```
Comment (cpcloud) [inline ibis/common/tests/test_grounds.py:346]: Can you move the cache tests into `tests/test_caching.py`? Ideally we closely match the module structure of the parent.

**Review by cpcloud — CHANGES_REQUESTED** [review]
Looks very good. Mostly just want to move the cache tests to their own file and uncomment the commented out tests (or remove them and a quick blurb about why in the PR).

File: ibis/backends/pandas/execution/join.py [hunk 5]
```diff
@@ -138,10 +138,10 @@ def execute_join(op, left, right, predicates, **kwargs):
     pd.DataFrame,
     pd.DataFrame,
     tuple,
-    tuple,
     (pd.Timedelta, type(None)),
+    tuple,
 )
-def execute_asof_join(op, left, right, predicates, by, tolerance, **kwargs):
+def execute_asof_join(op, left, right, by, tolerance, predicates, **kwargs):
```
[the remark this answers is not in the document]
Comment (kszucs) [reply ibis/backends/pandas/execution/join.py:144]: Yes, see https://github.com/ibis-project/ibis/pull/3650/files#diff-1ba248744f86eb25b4e863648c19353294f34c47b637d1fe88d66eccfc1e1119R100-R108

The signature of an annotable instance now follows:
`inherited_args + new_args + new_kwargs + inherited_kwargs`

File: ibis/tests/expr/test_value_exprs.py (hunk 3 above)
Comment (kszucs) [reply ibis/tests/expr/test_value_exprs.py:249]: Sure, it's a leftover.

File: ibis/common/grounds.py (hunk 2 above)
Comment (kszucs) [reply ibis/common/grounds.py:85]: Not sure since we alter the params mapping below.

File: ibis/common/tests/test_grounds.py (hunk 4 above)
Comment (kszucs) [reply ibis/common/tests/test_grounds.py:346]: This tests `Comparable` which is implemented in `grounds.py`

Comment (cpcloud) [reply ibis/common/tests/test_grounds.py:346]: Ah, ok. Fine as is then!

File: ibis/common/grounds.py (hunk 2 above)
Comment (cpcloud) [reply ibis/common/grounds.py:85]: Ah, right. Thanks!

**Review by cpcloud — APPROVED** [review]

**Pull request merged**